### ISSUE TITLE
NT8 broker execution issues with OnOrderUpate and OnExecutionUpdate - Rithmic and Interactive Brokers

### ISSUE THREAD

**Issue by MicroTrendsTom** [MicroTrendsLtd/NinjaTrader8#45]
Feedback received with thanks and Reported by Sidlercom
https://github.com/MicroTrendsLtd/NinjaTrader8/discussions/41#discussioncomment-290550

The crux is that NT8 is not 100% compatible or broker neutral for rithmic and Interactive brokers from within a strategy. 
OnOrderUpdate and OnExecutionUpdate and OnPoisitonUpdate may or may not be called in sequence or if at all.
Broker centric workarounds are required to make NT8 it work explicitly in the derived strategy classes instead of being handled in the plumbing below internally within NT8.

This is a debug trace supplied by Sidlercom

13.05.2021 19:44:39:177 > SidiSystemBase > OnStrategyTradeWorkFlowUpdated > GoLongSubmitOrderWorking
13.05.2021 19:44:39:177 > SidiSystemBase > OnOrderUpdate > Flat > ↑LE.MKT#3.ST | state=Working | limitPrice=0 | stopPrice=0 | qty=10 | filled=0 |avgPrice= 0 | time=13.05.2021 19:44:39 | Id=1193980675 | error=NoError
13.05.2021 19:44:39:187 > SidiSystemBase > OnOrderUpdate > Flat > ↑LE.MKT#3.ST | state=PartFilled | limitPrice=0 | stopPrice=0 | qty=10 | filled=1 |avgPrice= 13017.5 | time=13.05.2021 19:44:39 | Id=1193980675 | error=NoError
13.05.2021 19:44:39:187 > SidiSystemBase > OnExecutionUpdate > Long > ↑LE.MKT#3.ST > execution='123673936|1193980675|1193980675' instrument='MNQ 06-21' account='' exchange=Globex price=13017.5 quantity=1 marketPosition=Long orderId='1193980675' time='2021-05-13 19:44:39' sod=False statementDate='2021-05-13'
13.05.2021 19:44:39:188 > SidiSystemBase > OnExecutionUpdate > Long > ↑LE.MKT#3.ST > execution='123673937|1193980675|1193980675' instrument='MNQ 06-21' account='' exchange=Globex price=13017.75 quantity=6 marketPosition=Long orderId='1193980675' time='2021-05-13 19:44:39' sod=False statementDate='2021-05-13'
13.05.2021 19:44:39:188 > SidiSystemBase > OnOrderUpdate > Long > ↑LE.MKT#3.ST | state=PartFilled | limitPrice=0 | stopPrice=0 | qty=10 | filled=7 |avgPrice= 13017.71428571 | time=13.05.2021 19:44:39 | Id=1193980675 | error=NoError

partiall fill

13.05.2021 19:44:39:189 > SidiSystemBase > OnExecutionUpdate > Long > ↑LE.MKT#3.ST > execution='123673938|1193980675|1193980675' instrument='MNQ 06-21' account='' exchange=Globex price=13018 quantity=3 marketPosition=Long orderId='1193980675' time='2021-05-13 19:44:39' sod=False statementDate='2021-05-13'
13.05.2021 19:44:39:265 > SidiSystemBase > OnOrderUpdate > Long > ↑LE.MKT#3.ST | state=Filled | limitPrice=0 | stopPrice=0 | qty=10 | filled=10 |avgPrice= 13017.8 | time=13.05.2021 19:44:39 | Id=1193980675 | error=NoError

it doesn't go on here because Rithmic no longer calls the OnExecutionUpdate()

**Initial Solution  draft**
Pattern be implemented within the ATSQuadroStrategyBase that can handle these scenarios so the strategy is broker neutral and works the same for any connection etc Order Routing etc

It could be it detects Interactive Brokers and Rithmic and handles workflow processing
or a generic solution will one with the option for a parameter on/off to engage a monitor to check back on the state of the execution after an interval and then move the workflow forwards  - if CEP event driven model fails to move it forwards. 

**Risk mitigation with NT8 we are best to avoid multithreading and timers - due to deadlocks etc** 
A reliable pattern is to use events to drive the workflow forwards and use the tick/market flow to drive a virtual monitoring/timer interval/timeout  to check on the order/execution state.. This kind of pattern is used within the system already and so it should be easy to implement without risk of breaking the system.

In summary  if events dry up or are missed due to connectivity the market data will call a method to call back into the workflow to check the executions are filled and then move forwards.

https://github.com/MicroTrendsLtd/NinjaTrader8/blob/main/ATSQuadroStrategyBase/NinjaTrader%208/bin/Custom/Strategies/AlgoSystemBase.cs

For partial fills
Checks OnOrderUpdate also to fire OnExecutionUpdate 

protected override void OnMarketData(MarketDataEventArgs marketDataUpdate)  
is used to check back in after an interval
//call back in to workflow  - Monitor for Workflow events or for Error State
Assessment of workflow changes and when to set the call back to workflow pattern.

**Comment by jmscraig** [comment]
HI Tom, 

Saw this issue posted.   Just wanted to check if you have seen the significant work NT_Jim put in to address at least some of these issues.   Reply if not and I will do find some links to those conversations and sample templates.

James

**Comment by jmscraig** [comment]
I did not quickly find links to those conversations but did locate a key sample file I was looking to post.

In the attached .cs file by NT team member NT_Jim ...

.. from within OnMarketUPdate()  calls a function "private void InlineExecutionUpdate(Order order, double AvgFillPrice, int quantity, MarketPosition marketPosition, string orderId)" to act as a pseudo OnExectutionUpdate() event manager and attempts to address at least some of the same concerns you list above.

Note the attached file is NOT an NT8 export file.  I just zipped the *.cs file in place so that GitHub would let me attach it to this message.

[SampleUnmanagedRithmicInteractiveBrokersNonNTOnOrderUpdate2020.zip](https://github.com/MicroTrendsLtd/NinjaTrader8/files/6506135/SampleUnmanagedRithmicInteractiveBrokersNonNTOnOrderUpdate2020.zip)

`

**Comment by MicroTrendsTom** [comment]
> I did not quickly find links to those conversations but did locate a key sample file I was looking to post.
> 
> In the attached .cs file by NT team member NT_Jim ...
> 
> .. from within OnMarketUPdate() calls a function "private void InlineExecutionUpdate(Order order, double AvgFillPrice, int quantity, MarketPosition marketPosition, string orderId)" to act as a pseudo OnExectutionUpdate() event manager and attempts to address at least some of the same concerns you list above.
> 
> Note the attached file is NOT an NT8 export file. I just zipped the *.cs file in place so that GitHub would let me attach it to this message.
> 
> [SampleUnmanagedRithmicInteractiveBrokersNonNTOnOrderUpdate2020.zip](https://github.com/MicroTrendsLtd/NinjaTrader8/files/6506135/SampleUnmanagedRithmicInteractiveBrokersNonNTOnOrderUpdate2020.zip)
> 
> `
awesome no i have not even looked  sounds like i was thinking along the same lines a psuedo onExec  - excellent  i will have to try find time i hope in the next few weeks to get into this

**Comment by MicroTrendsTom** [comment]
> HI Tom,
> 
> Saw this issue posted. Just wanted to check if you have seen the significant work NT_Jim put in to address at least some of these issues. Reply if not and I will do find some links to those conversations and sample templates.
> 
> James

i hadnt even looked to be honest but i will now you have posted the example excellent

**Comment by MicroTrendsTom** [comment]
> I did not quickly find links to those conversations but did locate a key sample file I was looking to post.
> 
> In the attached .cs file by NT team member NT_Jim ...
> 
> .. from within OnMarketUPdate() calls a function "private void InlineExecutionUpdate(Order order, double AvgFillPrice, int quantity, MarketPosition marketPosition, string orderId)" to act as a pseudo OnExectutionUpdate() event manager and attempts to address at least some of the same concerns you list above.
> 
> Note the attached file is NOT an NT8 export file. I just zipped the *.cs file in place so that GitHub would let me attach it to this message.
> 
> [SampleUnmanagedRithmicInteractiveBrokersNonNTOnOrderUpdate2020.zip](https://github.com/MicroTrendsLtd/NinjaTrader8/files/6506135/SampleUnmanagedRithmicInteractiveBrokersNonNTOnOrderUpdate2020.zip)
> 
> `

AWESOME thank you -   sounds like i was thinking along the same lines a psuedo onExec  - excellent  i will have to try find time i hope in the next few weeks to get into this  - bogged down with asp.net core and blazor SignalR atm  to buikld a cloud based correlation gui and api for a product - ETA 2 weeks

**Comment by MicroTrendsTom** [comment]
Ok I had a brief look some caveats are
OnOrder doesnt always fire
Position tally cannot safely be done and left  within in scope OrderFills due to caveats external and internal

So we might look to quiz the connection or via parameter turn on a different workflow that would push through a call to onExec or other method to finalize the order enrty cycle. 
i would also like to examine the executions collection and quiz as to how it can be accurate if it does not  fir or trigger an update/add event  - so this could might be utilized to push through -  but care needs to be taken not to upset executions and trade calculations -   I have already seen NT8 completely  mis-report trades and executions with CQG Continuum  and so therefore have a different P&L stats on live trading versus live trading brokerage fills  - the whole thing is a bit delicate for sure.

**Comment by MicroTrendsTom** [comment]
Sorry but still far off from being able to address this myself due to other obligations kidnapping my time 24/7 almost coding for months on end etc

**Comment by MicroTrendsTom** [comment]
Unable to find time to investigate meanwhile waiting for  NT to fix this in the platform   as this is clearly just an omission their end and no doubt with no fiscal reasons to fix it as a priority.

Please note there are other projects as above that have fixed this issue by adding code to the engine - for all other broker users this wont be required etc 

Time permitting will wheel back to this if this projects becomes active as a dependency on other projects etc

**Comment by MicroTrendsTom** [comment]
One way to look at this is:
 This type of engine is only really good for backtesting and limited in live trading for attended and non fault tolerant commercial swing - but excellent for rapid algo design and testing.
 
After the algo design phase is complete in backtest,sim realtime it would be ready to add to a system that uses a realtime only engine and bypasses all the overheads and limitations of a NTY Strategy - such as the type we use here: https://algofuturestrader.zendesk.com/hc/en-us/articles/360055568611-NinjaTrader-7-AFT7-day-trading-futures-system-overview-of-components

have not migrated to NT8 due to lack of time - 3 years over due!...
But do have a server and cloud based system for NT8 that cant be brought to retail or open source -
looking for a way to do this... it might well be the bare bones and it gets its data points from a webapi et.... so we dont lose any really secret sauce etc...

**Comment by jmscraig** [comment]
So much in the market has shifted .. and a huge amount of liquidity has shifted to crypto that rather than lights-out auto-trading I am presently more interested in first in automation assisted trading, then leading toward human assisted auto-trading and finally some lights-out algo execution again.  

These images demonstrate the depth of complexity to be managed and in general what I am working to build with an eye toward crypto and classic futures for deployment.  https://algofuturestrader.zendesk.com/hc/en-us/articles/360055568611-NinjaTrader-7-AFT7-day-trading-futures-system-overview-of-components

If for no other reason than to respect the NinjaTrader Terms of Service for me NT8 needs to be an optional companion to the core platform.

For the interface stack I have been investigating 

1) non-NT8 WPF, WinUI 3 (including for the server side) and a few other Windows centric things. 

2) A few high speed bi-directional C# <> Javascript webhook interface scheme caught my eye:

2.1) One depends upon cref chromium, and another broad Javascript.   cref might be a little old but the author claims the integration is very light.. and can be free from asyc dependencies etc.     This would open up the possibility to leverage nice fast and responsive existing React and Vue UIs.     

2.2)  I saw another bi-directional C# <> Javascript webhook interface for integrating the slower-and-less-sexy-than-React Microsoft Blazor web UI ... another variant using Microsoft's not yet mature replacement for cref.   

3).  Regarding WinUI 3..  Third Party dependencies are a path that raises concerns for me but the three (WinUI3, Maui, Xamarin)  free forever libraries on this page have caught my eye.. they do simplify the work to build a UI..   https://www.devexpress.com/Products/Try/

Your thoughts?
James

**Comment by MicroTrendsTom** [comment]
im a  long  fan of devexpress and have an active annual sub
raw WPF in Visutal studio works great
NT8 WPF yes i try to avoid it as much as possible.

the only caveat is for things like  property girds if you need them the one in nt8 is not good to work with.
devexpress and others might be better or use a winforms host etc...

so this convo we can put into a discussion on uis  as this is about rithmic and so on...
point being im not going to resolve it as the best layer for a UI is not in a strategy as NT8 strategy is not the best thing for hybrid algo trading in NT8 - this will demonstrate exactly how soon  as i can drop onto it- http://algofuturestrader.com
